text/0000-custom-model-classes.md

```diff
@@ -92,9 +84,26 @@ class Store {
 
+## Change Notification
+
+In order for the record class to learn about changes to the underlying data, it will be passed a `NotificationManager` as a constructor argument, which will allow it to
+subscribe and unsubscribe to notifications of underlying changes to the data. Once `RecordData` calls one of the notification methods, the notification manager will call 
+any registered callback for the given identifier, and pass in the type of the notification, allowing the record the opportunity to repull the data if needed. There are no guarantees around the timing of the notification callback being called after `RecordData` informs the store of changes. We expect that in a modern Ember app with tracked properties, this wouldn't be the common path for tracking changes. 
+
+```ts
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload'): void;
+}
+    
+interface NotificationManager {
+  subscribe(identifier, NotificationCallback): void
+  unsubscribe(identifier, NotificationCallback): void
+ }
+```
+
 ## Exposing schema information
 
-We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer and the serializer/adapter layers.
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer, the serializer/adapter layers and the DebugAdapter. Schema methods support both static and dynamic schema computation. For static schemas, the method can always respond with a schema definition based on the type passed, and for dynamic changing schemas, it can look up the underlying data by the identifer which is passed in. 
```

Comment:
It is slightly concerning to me that we have to introduce schemas in this way here vs through a more complete design (which we want to achieve soon). I do wonder if we shouldn't have targeted designing a schema service first.

Reply:
That's why I designed them in a clunky way so we can replace them later and that most of the usage would be hidden by other public APIs. 

The only place you would have to manually deal with them would be in normalization methods. 
A medium term solution would be to not deprecate eachAttribute and eachRelationship being passed to normalize, and implement those in the shim class until a full featy

Reply:
I think shim class implementation is worth exploring